SSR: Add `wp-show` attribute directive processor

ockham opened the pull request:
Implement the `data-wp-show` directive, which is supposed to transform code like

```html
<div data-wp-show="falsyValue" class="my-class">
  <p>Children</p>
</div>
```

into

```html
<template data-wp-show="falsyValue">
  <div class="my-class">
    <p>Children</p>
  </div>
</template>
```

per [this discussion](https://github.com/WordPress/block-interactivity-experiments/issues/159#issuecomment-1458572890). (For truthy values of `data-wp-show`, it does not modify the HTML at all.)

Note that the behavior to move the `data-wp-show` attribute from the wrapped tag to the wrapping one was implemented in https://github.com/WordPress/block-interactivity-experiments/pull/141/commits/da195d73f469e7d9b39e228513eeeebab99d5b5e. If we decide that we'd rather keep the attribute on the wrapped tag, we can simply revert that commit.

gziolo commented on phpunit/directives/tags/wp-show.php:
```diff
@@ -0,0 +1,64 @@
+		$tags->next_tag();
+
+		$context_before = new WP_Directive_Context( array( 'myblock' => array( 'open' => true ) ) );
+		$context = $context_before;
```
Should it get cloned here? Otherwise, it’s the same object referenced.
```suggestion
		$context = clone $context_before;
```

or maybe the initial context should get assigned to the variable and used in the assertion later.

ockham replied:
Ah, good spot! 😅 I have a few more of those in other unit tests, will file a PR to update them.

gziolo reviewed (commented):
All those directives are neat!

ockham commented:
Per https://github.com/WordPress/block-hydration-experiments/issues/152#issuecomment-1448482028, I'll update this PR to turn the `wp-show` tag directive into an attribute directive.

ockham committed:
SSR: Add wp-show attribute directive processor

ockham commented:
I'm considering removing the `WP_Directive_Processor` I've added here, and rebasing this PR on #158 instead to use that PR's `WP_Directive_Processor` (yep, same name). This is currently blocked ([see](https://github.com/WordPress/block-hydration-experiments/pull/158#issuecomment-1452437543)) but should stop being a problem next week.

We could then rewrite the directive like so:

```diff
diff --git a/src/directives/attributes/wp-show.php b/src/directives/attributes/wp-show.php
index 89ed612..82b28cc 100644
--- a/src/directives/attributes/wp-show.php
+++ b/src/directives/attributes/wp-show.php
@@ -3,25 +3,19 @@
 require_once __DIR__ . '/../utils.php';
 
 function process_wp_show( $tags, $context ) {
-	if ( ! $tags->is_tag_closer() ) { // TODO: Exclude void and self-closing!
-		set_bookmark_for_directive( $tags, 'wp-show' );
+	if ( $tags->is_tag_closer() ) {
 		return;
 	}
 
-	$end = 'wp-show-closer';
-	$tags->set_bookmark( 'wp-show-closer' );
-	$start = seek_bookmark_for_directive( $tags, 'wp-show' );
-
 	$value = $tags->get_attribute( 'wp-show' );
-	if ( null !== $value ) {
-		$show = evaluate( $value, $context->get_context() );
+	if ( null === $value ) {
+		return;
+	}
 
-		if ( ! $show ) {
-			$content = $tags->get_content_inside_bookmarks( $start, $end );
-			$tags->set_content_inside_bookmarks( $start, $end, '<template>' . $content . '</template>' );
-		}
+	$show = evaluate( $value, $context->get_context() );
+	if ( ! $show ) {
+		$content = $tags->get_inner_html();
+		echo $content;
+		$tags->set_inner_html( '<template>' . $content . '</template>' );
 	}
-	$tags->seek( $end );
-	$tags->release_bookmark( $start );
-	$tags->release_bookmark( $end );
 }

```

dmsnell commented:
is there any foreseeable need for `set_outer_html()`? from what @adamziel discovered it seems like that one creates problems with the tag processor, but `set_inner_html()` doesn't. if we don't have to solve that problem I'd be in favor of letting it sit unsolved.

luisherranz commented:
Not having `set_outter_html` would limit our capability of wrapping the whole tag with `<template>` in `wp-show`:

From this:

```html
<div data-wp-show="false" class="my-class">
  <p>Children</p>
</div>
```

Doing this:

```php
$p->set_outter_html( '<template>' . $p->get_outter_html() . '</template>' );
```

```html
<template>
  <div data-wp-show="false" class="my-class">
    <p>Children</p>
  </div>
</template>
```

Instead of this:

```php
$p->set_inner_html( '<template>' . $p->get_inner_html() . '</template>' );
```

```html
<div data-wp-show="false" class="my-class">
  <template>
    <p>Children</p>
  </template>
</div>
```

We haven't decided about it yet, but yesterday [I shared an example](https://github.com/WordPress/block-hydration-experiments/issues/159#issuecomment-1452377597) of something that (I think) would not be possible with the `set_inner_html` option.

- https://github.com/WordPress/block-hydration-experiments/issues/159

This case could work with a "wrap tag" API as well, if that would make things easier.

---

It would also leave some leftovers when using `wp-fill`.

For example, this code:

```html
<table wp-slot="data"></table>

<template wp-fill="data">
  <tr>
    <td wp-text="id"></td>
    <td wp-text="firstName"></td>
    <td wp-text="lastName"></td>
  </tr>
</template>
```

Would end up as this using `$p->set_inner_html('')` to remove the fill once it has been moved to the correct place:

```html
<table wp-slot="data">
  <tr>
    <td wp-text="id"></td>
    <td wp-text="firstName"></td>
    <td wp-text="lastName"></td>
  </tr>
</table>

<template wp-fill="data"></template>
```

Leaving the empty template hanging around.

This case could work with a "remove tag" API as well, if that would make things easier.

---

The rest of the core directives we have in mind right now don't seem to need anything like similar to `set_outter_html`.

SantosGuillamot commented:
> This case could work with a "wrap tag" API as well, if that would make things easier.

Not sure if you meant this, but I suggested a [new approach](https://github.com/WordPress/block-hydration-experiments/issues/159#issuecomment-1453509805) to handle `wp-show` that wouldn't require `set_outter_html` if I am not mistaken: Restrict the `wp-show` directive to be used inside `<template>`. The same way you are using `wp-fill` in your examples. As mentioned there, I am not convinced about it, but I think you'll know better than me if it makes sense or not.

luisherranz commented:
It makes sense indeed. I've answered you on the other issue:

- https://github.com/WordPress/block-hydration-experiments/issues/159#issuecomment-1458572890

ockham commented:
Rebased on #158.

ockham committed:
Simplify implementation of wp-show

ockham commented:
We've [decided](https://github.com/WordPress/block-interactivity-experiments/issues/159#issuecomment-1458572890) that we'd like `wp-show` to work as follows:

> When `wp-show` is false:
> - Wrap it with `<template>` in the server (maybe also move the `wp-show` to the template).
> - Remove it from the DOM in the client.
>
> In the client, having the DOM elements wrapped by `<template>` is not useful. I mean, we already have the static virtual nodes that represent that HTML in memory, we don't need them in the DOM anymore.

I.e. we'd like the server-side rendered `wp-show` directive to return something like this:

```html
<template>
  <div data-wp-show="false" class="my-class">
    <p>Children</p>
  </div>
</template>
```

(with `data-wp-show="false"` possibly moved to `<template>`)

Our directive processor (#169) currently doesn't support `set_outer_html`, and it's been found that the latter could be somewhat tricky to implement.

As a consequence, I'd like to try implementing a `wrap_in_tag` (or so) method first, which should hopefully be easier, since it's not as potentially "destructive" to the current tag as `set_outer_html`.

ockham committed:
Implement basic wrap_in_tag method()

ockham committed:
Format wp-show

ockham committed:
Use wrap_in_tag to implement wp-show

ockham commented:
> I'd like to try implementing a `wrap_in_tag` (or so) method first, which should hopefully be easier, since it's not as potentially "destructive" to the current tag as `set_outer_html`.

Done with @michalczaplinski in https://github.com/WordPress/block-interactivity-experiments/pull/141/commits/39b2e1b0ab25982475d0eb47e822377ac7be2762 (but needs more polish 🇵🇱). Updated `wp-show` to use it in https://github.com/WordPress/block-interactivity-experiments/pull/141/commits/d3b36fa30a2066a67d7e097a5f33bd328b669877.

ockham committed:
Add data- prefix

ockham committed:
Format

ockham committed:
Add PHPDoc

ockham committed:
wrap_in_tag: Cover void tags

ockham committed:
wp-show: Add test to cover void tags

ockham committed:
Remove now-obsolete TODO comments

ockham committed:
Add test to demonstrate issue

ockham committed:
Return to original position after replacing

ockham committed:
wrap_in_tag: Bail early if we're at a tag closer

ockham committed:
Simply insert start and end tag in correct positions

ockham committed:
Document that the pointer keeps pointing at the same tag

ockham committed:
Simplify unit test

ockham committed:
Move attribute to wrapping tag

ockham commented on src/directives/class-wp-directive-processor.php:
```diff
@@ -123,6 +123,75 @@ public function get_balanced_tag_bookmarks() {
+	 * Note that the internal pointer will continue to point to the same tag as before
+	 * calling the function.
+	 *
+	 * @param string $tag An HTML tag, specified in uppercase (e.g. "DIV").
+	 * @return string|false The name of a bookmark pointing to the wrapping tag opener
+	 *                      if successful; false otherwise.
+	 *
+	 * @todo Allow passing in tags with attributes, e.g. <template id="abc">?
```
Note that we're only accepting an uppercase HTML tag such as `DIV` or `TEMPLATE` as the `$tag` argument. We don't accept it wrapped in angle brackets, or with attributes (`<template attr="value>`), as that would require us to apply HTML parsing to the `$tag` argument itself -- and would allow for many more edge cases for which we'd need to decide how to behave (E.g. what if we pass two tags (`<template><div>`) -- do we wrap in both? What if one of them is void (`<template><img>`)? What if there's a basic syntax error (`<template`)?)

We can eschew most of these questions by only allowing a tag name; handling of attributes can be done after the fact and separately in `process_wp_show`.

dmsnell replied:
maybe we could call the argument `$tag_name` to communicate that better? also I think we discussed this, but the part about "in uppercase" shouldn't matter, right?

ockham committed:
Fail if trying to wrap in void tag

ockham commented:
Hmm, weird 🤔 It's definitely possible that the directive isn't quite working correctly (as this one requires some things that we didn't previously need in other directives, such as moving the internal pointer around).

However, I couldn't reproduce when modifying the unit test:

```diff
diff --git a/phpunit/directives/attributes/wp-show.php b/phpunit/directives/attributes/wp-show.php
index f9314c1..3f812f5 100644
--- a/phpunit/directives/attributes/wp-show.php
+++ b/phpunit/directives/attributes/wp-show.php
@@ -34,7 +34,7 @@ class Tests_Directives_WpShow extends WP_UnitTestCase {
        }
 
        public function test_directive_wraps_content_in_template_if_when_is_false() {
-               $markup = '<div data-wp-show="context.myblock.open">I should not be shown!</div>';
+               $markup = '<div data-wp-show="context.myblock.open"><span>I should not be shown!</span></div>';
 
                $tags = new WP_Directive_Processor( $markup );
                $tags->next_tag();
@@ -46,7 +46,7 @@ class Tests_Directives_WpShow extends WP_UnitTestCase {
                $tags->next_tag( array( 'tag_closers' => 'visit' ) );
                process_wp_show( $tags, $context );
 
-               $updated_markup = '<template data-wp-show="context.myblock.open"><div >I should not be shown!</div></template>';
+               $updated_markup = '<template data-wp-show="context.myblock.open"><div ><span>I should not be shown!</span></div></template>';
 
                $this->assertSame( $updated_markup, $tags->get_updated_html() );
                $this->assertSame( $context_before->get_context(), $context->get_context(), 'data-wp-show directive changed context' );
```

I also tried adding more 

```
		$tags->next_tag( array( 'tag_closers' => 'visit' ) );
		process_wp_show( $tags, $context );
```

pairs to emulate iterating through the entire HTML, but that also didn't cause the test to fail 🤔 

I'll have another look next week! Hopefully we can find a minimal test case that we can turn into a unit test to guard against this kind of error 🤞

SantosGuillamot commented:
I still have to review the code properly, but I'm able to reproduce the issue with the following code:

```php
function add_wp_show($content)
{
	return $content . "<div data-wp-context='{ \"isOpen\": false }'><div data-wp-show='context.isOpen'><span>WP Show</span></div></div>";
}
add_filter(
	'render_block_core/post-title',
	'add_wp_show',
	10,
	1
);
```

ockham commented:
> I'll try to add a basic unit test to cover this and see if it fails 😊

It's a bit tricky to distill this into a minimal test case -- still working on it.

@SantosGuillamot In the meantime, could you try if this patch (applied on top of this PR's branch) fixes the issue?

```diff
diff --git a/wp-directives.php b/wp-directives.php
index e4be9f3..5187b2a 100644
--- a/wp-directives.php
+++ b/wp-directives.php
@@ -230,7 +230,7 @@ function process_directives_in_block( $block_content ) {
        return $tags->get_updated_html();
 }
 add_filter(
-       'render_block',
+       'the_content',
        'process_directives_in_block',
        10,
        1
```

SantosGuillamot commented:
> @SantosGuillamot In the meantime, could you try if this patch (applied on top of this PR's branch) fixes the issue?

I quickly tested it, and it doesn't seem to work, but I believe `the_content` just returns the Post content, and I'm using most of the blocks outside of the Post content (in a template). I can try tomorrow to include those blocks in the post content itself.

ockham commented:
> I quickly tested it, and it doesn't seem to work, but I believe `the_content` just returns the Post content, and I'm using most of the blocks outside of the Post content (in a template). I can try tomorrow to include those blocks in the post content itself.

Thank you! Yeah, that makes a lot of sense.

I'll need to keep looking for a better filter to use (that includes templates and content); or alternatively to continue doing it on a per-block basis (which might also have some advantages), in which case I'd need to find a way to avoid processing directives twice for nested blocks 🤔

dmsnell commented:
@ockham this is something we should talk about this upcoming week, but I wonder how bad it would be to simply add a new attribute to prevent double-processing.

```php
$p->set_attribute( 'data-wp-has-processed', true );
```

Then we can bail if we've already set that.

```php
if ( $p->get_attribute( 'data-wp-has-processed' ) ) {
    continue;
}
```

in any case this might be a reasonable method to continue exploring until a clearly better solution is found, and this would mean we don't have to rely on people _doing it right_ by calling this in the appropriate places and not in others. I suspect there's an open list of places someone might call the processor multiple times that we don't anticipate.

luisherranz commented:
> someone might call the processor multiple times

I don't think devs should be manually calling the directives processor at all. It should be called only by Gutenberg.

ockham committed:
Increase test coverage -- new test currently fails

ockham committed:
Don't process directives in inner blocks

ockham committed:
Add test case to cover issue

ockham committed:
Fix test

ockham committed:
More minimal

ockham committed:
Add another failing test case

ockham committed:
Flush before moving to next tag

ockham committed:
We might not want to limit to interactive blocks only

ockham commented:
I think that #215 should finally unblock this PR 🤞

ockham committed:
Revert "We might not want to limit to interactive blocks only"

This reverts commit b7c69257006d71e0c601c2b0b35be9fd2f74299c.

ockham committed:
Revert "Don't process directives in inner blocks"

This reverts commit 16faac9f07967020b42267f4a85d36cae36fdb2c.

ockham committed:
Fix test

ockham commented on phpunit/directives/wp-directive-processor.php:
```diff
@@ -125,4 +125,55 @@ public function test_set_inner_html_invalidates_bookmarks_that_point_to_replaced
+		$tags->next_tag( 'section' );
+		$tags->wrap_in_tag( 'TEMPLATE' );
+		$this->assertSame( '<div class="foo">outside</div><template><section><div><img>inside</div></section></template>', $tags->get_updated_html() );
+		$this->assertSame( 'SECTION', $tags->get_tag() );
```
Note that this is currently [failing](https://github.com/WordPress/block-interactivity-experiments/actions/runs/4733465266/jobs/8401035058?pr=141) (but fixed by #215). The error is:

```
2) WP_Directive_Processor_Test::test_wrap_in_tag_wraps_balanced_tag_correctly_when_called_after_add_class
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'SECTION'
+'TEMPLATE'

/var/www/html/wp-content/plugins/block-interactivity-experiments/phpunit/directives/wp-directive-processor.php:147
```

Funny enough, if we change the `foo` to `fooooooo`, the error is different:

```
2) WP_Directive_Processor_Test::test_wrap_in_tag_wraps_balanced_tag_correctly_when_called_after_add_class
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'SECTION'
+'DIV'

/var/www/html/wp-content/plugins/block-interactivity-experiments/phpunit/directives/wp-directive-processor.php:147
```

(I.e. we're now at the `<div>` rather than the `<section>`.)

dmsnell commented on src/directives/class-wp-directive-processor.php:
```diff
@@ -131,6 +131,84 @@ public function get_balanced_tag_bookmarks() {
+			return false;
+		}
+
+		$this->get_updated_html(); // Apply potential previous updates.
```
After applying the fix in ~dmsnell/wordpress-develop#8~ WordPress/wordpress-develop#4345 and removing this call to `get_updated_html()` my test case, where I reproduced at least some of the errors we found in this PR, succeeded.

dmsnell replied:
And this is the test I used to reproduce…

```php
public function blahblahblah() {
	$p = new WP_Directive_Processor( '<div><img hidden></div>' );

	$p->next_tag();
	$p->next_tag('img');
	$p->set_attribute( 'inert', true );
	$p->remove_attribute( 'hidden' );
	$b = $p->wrap_in_tag( 'figure' );
	$this->assertSame( 'IMG', $p->get_tag() );
	$p->seek( $b );
	$this->assertSame( 'FIGURE', $p->get_tag() );
	$p->next_tag();
	$this->assertSame( 'IMG', $p->get_tag() );

	$this->assertSame( '<div><figure><img inert ></figure></div>', $p->get_updated_html() );
}

```

ockham replied:
Thank you!

I just tested your [patch](https://github.com/WordPress/wordpress-develop/pull/4345) (plus the removal of that `get_updated_html()` call inside `wrap_in_tag()`, and my changes from #215) against the Movies Demo. Unfortunately, the [issue I found there](https://github.com/WordPress/block-interactivity-experiments/pull/215#issuecomment-1513162325) still persists 😕 

On a sidenote -- it strikes me as odd that we need to remove a `get_updated_html()` call. Since I've conceptually thought of it as a flushing method, I was silently assuming that there was no harm in calling it. Am I mistaken here? Or is there still something wrong with it that we need to fix?

ockham replied:
Also, the tests that are currently failing on this branch are still failing with your patch applied and that `get_updated_html` call removed 😕 

To verify that your patch was properly picked up, I copied the test from https://github.com/WordPress/wordpress-develop/pull/4345 into `phpunit/directives/wp-directive-processor.php` and ran unit tests. That added test passed (which it didn't when I un-applied your patch).

dmsnell replied:
> it strikes me as odd that we need to remove a get_updated_html() call

yeah I've been working on figuring this out better

> still failing with your patch applied

this is why I had tried to carefully word that my patch fixed _my tests_ as I suspected we could still have more in this PR.

ockham replied:
For the record, I've filed a [PR](https://github.com/WordPress/wordpress-develop/pull/4355) against `wordpress-develop` based on this one to make it a bit easier to reproduce and test fixes.

ockham replied:
Noting that @dmsnell managed to reduce that test case to be more minimal and has filed a PR with a tentative fix: https://github.com/WordPress/wordpress-develop/pull/4371

dmsnell replied:
and it's fixed now!

ockham commented:
Here's an update and recap of the issue(s) with `wp-show`: https://github.com/WordPress/block-interactivity-experiments/issues/120#issuecomment-1519712031

ockham committed:
Rewrite wp-show processor

ockham committed:
Blunt but effective

ockham committed:
Remove now-obsolete unit tests

ockham committed:
PHPDoc

ockham commented:
I've just removed a few unit tests that I introduced to help me pin-point the issue. The good news is that @dmsnell has filed a [PR](https://github.com/WordPress/wordpress-develop/pull/4371) against `wordpress-develop` that will fix them 🎉 

While testing that PR, I've noticed that its effect on `wp-show` seems to be equivalent to https://github.com/WordPress/block-interactivity-experiments/pull/141/commits/aad67cc7290f7ff9ad8a1f7f717b6cee56ea8063. Furthermore, I noticed that it seems to be possible to fix the [remaining issue](https://github.com/WordPress/block-interactivity-experiments/pull/215#issuecomment-1513162325) that I encountered with [this commit](https://github.com/WordPress/block-interactivity-experiments/pull/141/commits/3dbff3f2696b9f2e00cacd5dafb9bd40ba298c29).

This means that I hope that I've managed to finally unblock this PR 🤞 So it's ready for another round of reviews!

gziolo commented:
Excellent work on this PR. All the unit tests included make catching up with the implementation details easier. What's the reason for removing `test_directive_wraps_complex_content_in_template_if_when_is_false` in https://github.com/WordPress/block-interactivity-experiments/pull/141/commits/02a4a5be284841e74259ab5d6409511fb7cb6a2c? I like how it illustrates the more complex usage of the directive (excluding `process_wp_class` part). 

Have you explored how nesting of the same directive would work in practice? Example:

```html
<div data-wp-show="context.myBlock.open">
    I should be shown!
    <div data-wp-show="context.myOtherBlock.open">I should be shown!</div>
    <div data-wp-show="!context.myOtherBlock.open">I should not be shown!</div>
</div>
```

vs

```html
<div data-wp-show="context.myBlock.open">
    I should not be shown!
    <div data-wp-show="context.myOtherBlock.open">I should be shown!</div>
    <div data-wp-show="!context.myOtherBlock.open">I should not be shown!</div>
</div>
```

ockham committed:
Add unit test for nested directives

ockham commented:
> Have you explored how nesting of the same directive would work in practice? Example:
> 
> ```
> <div data-wp-show="context.myBlock.open">
>     I should be shown!
>     <div data-wp-show="context.myOtherBlock.open">I should be shown!</div>
>     <div data-wp-show="!context.myOtherBlock.open">I should not be shown!</div>
> </div>
> ```
> 
> vs
> 
> ```
> <div data-wp-show="context.myBlock.open">
>     I should not be shown!
>     <div data-wp-show="context.myOtherBlock.open">I should be shown!</div>
>     <div data-wp-show="!context.myOtherBlock.open">I should not be shown!</div>
> </div>
> ```

Added in https://github.com/WordPress/block-interactivity-experiments/pull/141/commits/35d7c99753a05e4b8ec58ac22ace03fd8bf0431b. @gziolo LMK if this covers what you meant or if you'd like something more or different! 😊

gziolo commented on phpunit/directives/attributes/wp-show.php:
```diff
@@ -0,0 +1,176 @@
+			<template data-wp-show="context.myBlock.open"><div >
+				I should be shown!
+				<div data-wp-show="context.myOtherBlock.open">I should be shown!</div>
+				<template data-wp-show="!context.myOtherBlock.open"><div >I should not be shown!</div></template>
```
Should it convert `div` to `template` inside another `template`? That was the reason I wanted to see those tests, so they cover edge cases like that.

gziolo commented on phpunit/directives/attributes/wp-show.php:
```diff
@@ -0,0 +1,176 @@
+	public function test_nested_directives_within_wp_show_with_falsy_value() {
+		$markup = <<<END
+			<div data-wp-show="context.myBlock.open">
+				I should be shown!
```
Nitpick: when using the falsy value, it should follow other tests and the text used:
```suggestion
				I should not be shown!
```

gziolo replied:
The expected result needs to get updated, too. Otherwise the test fails.

ockham committed:
Fix test string

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

gziolo commented on phpunit/directives/attributes/wp-show.php:
```diff
@@ -0,0 +1,176 @@
+
+		$updated_markup = <<<END
+			<template data-wp-show="context.myBlock.open"><div >
+				I should be shown!
```
We need it now to make the test pass:
```suggestion
				I should not be shown!
```

ockham replied:
😅

ockham committed:
Update tests accordingly

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

gziolo reviewed (approved):
I don't know the codebase, so take my feedback with a grain of salt. I'm approving this PR to show that the current testing coverage proves that the basic functionality is fully functional. There is one edge case that I highlighted in https://github.com/WordPress/block-interactivity-experiments/pull/141#discussion_r1184807429, which might be problematic, but we can test it only with the nested templated handled on the front end. It requires input from someone more knowledgable in the area.

ockham commented on src/directives/wp-process-directives.php:
```diff
@@ -68,6 +68,7 @@ function wp_process_directives( $tags, $prefix, $directives ) {
 		foreach ( $attributes as $attribute ) {
 			call_user_func( $directives[ $attribute ], $tags, $context );
 		}
+		$tags->get_updated_html();
```
Since WP 6.2.1 -- which includes [this fix](https://github.com/WordPress/wordpress-develop/pull/4371) -- has now been released, we might be able to remove this line now (i.e. revert 42ab641c9a4dc9fc37e136eb596fc3eb288dd074).

ockham replied:
Seems to be working: https://github.com/WordPress/block-interactivity-experiments/pull/141/commits/9c95d2df881051447818730f0e46d1b432ea7ee2

ockham committed:
Revert "Flush before moving to next tag"

This reverts commit 9430be82eae23bc3be42f40c888d8dc68429f1f7.

luisherranz commented:
I'm going to close this issue as part of the migration to the Gutenberg repository.

This task was added to the Roadmap and we'll open a new issue/discussion once it's time to start working on this again.
